stdlib/buffer.gr

```diff
@@ -0,0 +1,463 @@
+  }
+}
+
+/**
+ * Returns the number of bytes currently contained in the buffer.
+ * @param buf: Buffer - The buffer
+ * @returns Number
+ */
+export let length = (buf) => buf.size
+
+/**
+ * Returns the number of bytes currently contained in the buffer.
+ * @param buf: Buffer - The buffer
+ * @returns Void
+ */
+export let clear = (buf) => {
+  Bytes.fill(0x0l, buf.data)
+  buf.size = 0
+}
+
+/**
+ * Empty the buffer and deallocate the internal byte sequence holding the buffer contents.
+ * @param buf: Buffer - The buffer
+ * @returns Void
+ */
+export let reset = (buf) => {
+  buf.data = Bytes.make(buf.initialSize)
+  buf.size = 0
+}
+
+/**
+ * Empty the buffer and deallocate the internal byte sequence holding the buffer contents.
+ * @param buf: Buffer - The buffer
+ * @returns Void
+ */
+export let truncate = (buf) => {
+  buf.data = Bytes.make(buf.initialSize)
+  buf.size = 0
+}
+
+/**
+ * Returns a copy of the current contents of the buffer as a Bytes
+ * @param buf: Buffer - The buffer
+ * @returns String
+ */
+export let toBytes = (buf) => {
+  let len = Bytes.length(buf.data)
+  if (buf.size == len) {
+    buf.data
+  } else {
+    Bytes.slice(0, buf.size, buf.data)
+  }
+}
+
+/**
+ * Returns a slice of the current contents of the buffer as a Bytes
+ * @param i: Number - start
+ * @param len: Number - length
+ * @param buf: Buffer - The buffer
+ * @returns String
+ */
+export let toBytesSlice = (i, len, buf) => {
+  checkIndexIsInBounds(i, len, buf)
+  Bytes.slice(i, len, buf.data)
+}
+
+/**
+ * Returns a copy of the current contents of the buffer as a String
+ * @param buf: Buffer - The buffer
+ * @returns String
+ */
+export let toString = (buf) => {
+  Bytes.toString(toBytes(buf))
+}
+
+/**
+ * Returns a copy of the current contents of the buffer as a String
+ * @param i: Number - start
+ * @param len: Number - length
+ * @param buf: Buffer - The buffer
+ * @returns String
+ */
+export let toSubString = (i, len, buf) => {
+  Bytes.toString(toBytesSlice(i, len, buf))
+}
+
+
+/**
+ * Copies a range of bytes from a source buffer to a given location in a destination buffer.
+ * @param srcPos: Number - The starting byte index to copy bytes from
+ * @param dstPos: Number - The starting byte index to copy bytes into
+ * @param len: Number - The amount of bytes to copy from the source buffer
+ * @param src: Bytes - The source buffer
+ * @param dst: Bytes - The destination buffer
+ * @returns Void
+ */
+export let move = (srcOff, dstOff, len, src, dst) => {
+  // TODO: bounds checks
+  Bytes.move(srcOff, dstOff, len, src.data, dst)
+}
+
+/**
+ * Appends the bytes of a Char to the end of the buffer
+ * @param char: Char - the char to end
+ * @param buf: Buffer - The buffer
+ * @returns Void
+ */
+export let addChar = (char, buf) => {
+  // TODO: Should I just @disableGC memcopy stuff in this function to improve perf 👀
+  let n = getCharAsWasmI32(char)
+  let byteLen = getCharByteLength(n)
+  let len = match (byteLen) {
+    1n => {
```

Comment:
@ospencer your thoughts on whether this should depend on `String.encode` (once it lands) or maybe vice versa?

Reply:
It's probably pretty close to the time that we figure out how we want to handle shared utilities in the stdlib. I'd like if they could both pull the same utility so neither has to depend on the other.